src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/LengthBucketsFrozenDictionary.cs

```diff
@@ -50,59 +48,59 @@ internal sealed class LengthBucketsFrozenDictionary<TValue> : FrozenDictionary<s
+                // to have a LOT of strings of different sizes.
                 return null;
             }
 
-            var lengthBuckets = new KeyValuePair<string, int>[spread][];
+            // Instead of creating a dictionary of lists or a multi-dimensional array
+            // we rent a single dimension array, where every bucket has five slots.
+            // The bucket starts at (key.Length - minLength) * 5.
+            // Each value is an index of the key from _keys array
+            // or just -1, which represents "null".
+            int[] buckets = ArrayPool<int>.Shared.Rent(arraySize);
+            buckets.AsSpan(0, arraySize).Fill(-1);
 
-            // Iterate through each bucket, filling the keys/values arrays, and creating a lookup array such that
-            // given a string length we can index into that array to find the array of string,int pairs: the string
-            // is the key and the int is the index into the keys/values array for the corresponding entry.
-            int index = 0;
-            foreach (KeyValuePair<int, List<KeyValuePair<string, TValue>>> group in groupedByLength)
+            for (int i = 0; i < keys.Length; i++)
             {
-                KeyValuePair<string, int>[] length = lengthBuckets[group.Key - minLength] = new KeyValuePair<string, int>[group.Value.Count];
-                int i = 0;
-                foreach (KeyValuePair<string, TValue> pair in group.Value)
-                {
-                    length[i] = new KeyValuePair<string, int>(pair.Key, index);
-                    keys[index] = pair.Key;
-                    values[index] = pair.Value;
+                string key = keys[i];
+
+                int index = (key.Length - minLength) * MaxPerLength;
 
-                    i++;
-                    index++;
+                if (buckets[index] < 0) buckets[index] = i;
+                else if (buckets[index + 1] < 0) buckets[index + 1] = i;
+                else if (buckets[index + 2] < 0) buckets[index + 2] = i;
+                else if (buckets[index + 3] < 0) buckets[index + 3] = i;
+                else if (buckets[index + 4] < 0) buckets[index + 4] = i;
+                else
+                {
+                    // If we've already hit the max per-bucket limit, bail.
+                    ArrayPool<int>.Shared.Return(buckets);
+                    return null;
                 }
             }
 
-            return new LengthBucketsFrozenDictionary<TValue>(keys, values, lengthBuckets, minLength, comparer);
+#if NET6_0_OR_GREATER
+            // We don't need an array with every value initialized to zero if we are just about to overwrite every value anyway.
+            // AllocateUninitializedArray is slower for small inputs, hence the size check.
+            int[] copy = arraySize < 1000
+                ? new int[arraySize]
+                : GC.AllocateUninitializedArray<int>(arraySize);
```

Comment:
The first thing `GC.AllocateUninitializedArray<int>` does is a length < 512 check.   Can we just let it do its thing rather than also checking here?